### ISSUE TITLE
Terminal mode should probably not send CSI-u sequences for shift+enter and shift+space

### ISSUE THREAD

**Issue by mortie** [neovim/neovim#24093]
### Problem

When used with a terminal app which supports CSI-u sequences, neovim will use those to get more fine-grained key combo information for key combos, to be able to detect sequences such as shift+enter or shift+space. In `:terminal` mode, nvim will reproduce those CSI-u sequences. That means, when you hit shift+space in `:terminal` mode, neovim will send `\e[32;2u` to the terminal instead of a space character, and when you hit shift+enter, neovim will send `\e[13;2u` instead of a line feed character.

This is problematic. On most systems I've encountered, `\e[32;2u` and `\e13;2u` aren't interpreted as space or enter by default by any shell, which means the user ends up with garbage in their terminal if they accidentally hit space before letting go of enter. While some programs can be manually configured to understand the CSI-u sequences properly, it seems unreasonable to expect the user to manually go through every system they ever interact with (directly, through SSH, through serial, whatever) and reconfigure readline and whatever other line reading libraries are in use by terminal software they use.

I believe neovim should, by default, do what iTerm2 does by default: send a space character when shift+space is pressed and a line feed character when shift+enter is pressed, unless CSI-u mode is enabled by a program printing `\e[>1u` to the terminal, at which point it starts reporting those sequences using CSI-u. If that's too much work, I think disabling CSI-u for shift+enter and shift+space would be better than the current situation.

Note: As a workaround, shift+space and shift+enter can be remapped to space and enter with `tmap <S-Space> <Space>` and `tmap <S-Enter> <Enter>`.

### Steps to reproduce

1. `nvim --clean`
2. `:terminal`
3. Press shift+enter or shift+space

### Expected behavior

The shift key should be ignored, a space or enter should be written to the terminal. This is what iTerm2 and, I believe, most other terminals with CSI-u support do.

### Neovim version (nvim -v)

0.9.1

### Vim (not Nvim) behaves the same?

No, vim's :terminal mode sends a space and enter character for shift+space/shift+enter

### Operating system/version

macOS 13.4

### Terminal name/version

3.4.19

### $TERM environment variable

xterm-256color

### Installation

Homebrew

**Comment by zeertzjq** [comment]
Search existing issues

**Comment by mortie** [comment]
With respect, I don't think this issue should be closed, and I don't think the others should have been either.

This seems to be the issue this is a duplicate of: https://github.com/neovim/neovim/issues/22168

That issue is closed as a duplicate of this: https://github.com/neovim/neovim/issues/21360, which is the same issue... but it's also closed with no comment other than it being a duplicate of https://github.com/neovim/neovim/issues/18735.

There are two problems here:

1. I don't understand why https://github.com/neovim/neovim/issues/18735 was closed. @justinmk closed it with the comment that neovim "can't be expected to force nested terminal programs to also support CSIu", which, while correct, is irrelevant. iTerm2 also supports CSIu, and iTerm2 also can't be expected to force programs to support CSIu... which is why iTerm2 doesn't send CSIu until the program running in the terminal has claimed to support CSIu by outputting `\e[>1u`. I'm suggesting that neovim should act the same way as other terminal emulators.
2. Even if neovim doesn't want to be well-behaved with regards to CSIu sequences in general, I think there's a vast difference between "ctrl-/ doesn't work in fzf", which is a pretty niche issue which could be reasonably fixed by making fzf support CSIu, and "basic typing is broken in all popular shells". I think a special workaround for shift+space is warranted even if fzf remains broken until it receives CSIu support.

I request that this issue is re-opened so that we can actually have these discussions. If not, I want the neovim team to express clearly that the intention is for shift+space to remain broken, without hiding behind this tangentially related fzf issue.

**Comment by zeertzjq** [comment]
Hmm, there problem here is that it is not possible to know if the program has requested CSI-u, as libvterm doesn't provide an API for that.

**Comment by justinmk** [comment]
The title sounds unconditional. Should it be conditional on whether the child process requests CSI-u ?

**Comment by krobelus** [comment]
yeah I think the terminal should go for compatibility by default, and if a child process explicitly requests CSI u then send that.

**Comment by allpan3** [comment]
Has something been done to this? In 0.10.3 `<shift-space>` indeed produces a space in every mode. `<shift-enter>` still sends `^[[13;2u`.

As someone who wants to map `<shift-space>`, this actually takes away the keybinding. 
Also, I'm confused as to whether neovim supports CSI-u in terminal mode or not. The discussion here seems to suggest it supports, but `<C-i>` still produces `<Tab>` in terminal mode instead of `^[[105;5u`.

**Comment by gpanders** [comment]
> The request in the issue title sounds unconditional. Should it be conditional on whether the child process requests CSI-u ?

This should be trivial to support now that we’ve vendored vterm.

**Comment by krobelus** [comment]
FWIW with the kitty keyboard protocol enabled, shift-space may be sent as space depdenting on the exact flags, see https://codeberg.org/dnkl/foot/issues/1642#issuecomment-1660461

**Comment by gpanders** [comment]
> > The request in the issue title sounds unconditional. Should it be conditional on whether the child process requests CSI-u ?
> 
> This should be trivial to support now that we’ve vendored vterm.

Ok it turns out it's not trivial at all. It's easy enough to disable CSIu encoding in vterm (it's enabled unconditionally by default), but supporting even level one of the kitty keyboard protocol ("Disambiguate escape codes") will not be straightforward because Nvim "merges" modifiers so that by the time the keys get processed, `<C-I>` is indistinguishable from `<Tab>` (even though the parent terminal emulator distinguishes these because Nvim enables the kitty keyboard protocol itself, Nvim then turns around and re-ambiguates it)[^1].

So even if an application running in a terminal buffer enables the disambiguate escape codes mode with `CSI > 1 u`, it won't make any difference right now. We'd need to make changes deep in Nvim's input processing code.

I think the simplest thing to do now is just disable vterm's unconditional CSIu encoding. That solves this issue at least.

[^1]: It doesn't do this if there is a mapping for `<C-I>` though

**Comment by zeertzjq** [comment]
Vim has a `no_reduce_keys` global variable. Nvim may need it in Terminal mode.

**Comment by zeertzjq** [comment]
Related: [Vim patch 8.1.2134](https://github.com/vim/vim/commit/6a0299d8f4c7a64c64d60a6bb39cfe6eaf892247)
